File: desktop-app/src/main/kotlin/scraper/gov_shared.kt
import io.netty.util.concurrent.Promise
import org.example.model.INews
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.chrome.ChromeOptions
import java.util.Date

fun gov_shared(n: Int, website: String): List<INews> {
    val news: MutableList<INews> = mutableListOf()

    val options = ChromeOptions()
    options.addArguments("--headless=new")
    val driver: WebDriver = ChromeDriver(options)
    driver.get(website)
    val elements = driver.findElements(By.cssSelector(".title a")).map { element -> element.getAttribute("href") }

    for (i in 0 until n) {
        val url = elements[i]

        if (url.isEmpty()) {
            println("Cannot get url")
            continue
        }

        val newsItem = getNewspage(driver, "$url")
        news.add(newsItem)

        if (i == n - 1) {
            break
        }
    }

    driver.quit()
    return news
}

fun getNewspage(driver: WebDriver, url: String): INews {
    try {
        driver.get(url)

        val authorElement = driver.findElement(By.cssSelector(".organisations a"))
        val author: String = authorElement.text.trim()
        if (author.isEmpty()) {
            println("Cannot fetch author from page: $url")
        }

        val contentElements = driver.findElements(By.cssSelector(".content.col.left.grid-col-8 > :not(:first-child)"))
        val content = contentElements.joinToString("\n") { it.text.trim() }
        if (content.isEmpty()) {
            println("Cannot fetch content from page: $url")
        }

        val titleElement = driver.findElement(By.cssSelector(".page-head h1"))
        val title: String = titleElement.text.trim()

        val dateElement = driver.findElement(By.cssSelector(".info time"))
        val dateUnparsed: String = dateElement.text.trim()
        if (dateUnparsed.isEmpty()) {
            println("Cannot get date")
        }
        val dateSplit = dateUnparsed.split(". ")
        val date = Date(
            dateSplit[2].toInt(),
            dateSplit[1].toInt() - 1,
            dateSplit[0].toInt() + 1,
            0,
            0,
            0
        )

        return INews(
            title.trim(),
            url,
            date,
            listOf(author),
            content,
            emptyList(),
            ""
        )
    } catch (error: Exception) {
        println("Cannot fetch page: $url")
        return INews("", url, Date(), emptyList(), "", emptyList(), "")
    }
}

File: server/scraper/websites/gov_shared.ts
import cheerio from 'cheerio';
import axios from 'axios';
import { INews } from '../../model/News';

async function gov_shared(n: number, website: string) {
  const news: INews[] = [];
  const news_promises: Promise<{ author: string; content: string }>[] = [];

  const response = await axios.get(website);
  const $ = cheerio.load(response.data);

  $('.list-item').each((i, element) => {
    const title: string | undefined = $(element).find('.title').text();
    if (!title) {
      console.log('Cannot get title');
      return false;
    }
    let url: string | undefined = $(element).find('a').attr('href');
    if (!url) {
      console.log('Cannot get url');
      return false;
    }

    url = `http://www.gov.si${url}`;

    const news_promise = get_newspage(url);
    news_promises.push(news_promise);

    const date_unparsed: string | undefined = $(element)
      .find('.datetime time')
      .text();
    if (!date_unparsed) {
      console.log('Cannot get date');
      return false;
    }

    const date_split = date_unparsed.split('. ');
    const date = new Date(
      +date_split[2],
      +date_split[1] - 1,
      +date_split[0] + 1,
      0,
      0,
      0
    );

    news.push({
      title: title.trim(),
      url: url,
      date: date,
      authors: [],
      content: '',
      categories: [],
      views: [],
      location: {
        type: 'Point',
        coordinates: [0, 0],
      },
    });

    if (i == n - 1) {
      return false;
    }
  });

  const news_responses = await Promise.all(news_promises);
  news_responses.forEach((news_response, i) => {
    news[i].authors.push(news_response.author);
    news[i].content = news_response.content;
  });

  return news;
}

async function get_newspage(
  url: string
): Promise<{ author: string; content: string }> {
  try {
    const news_response = await axios.get(url);
    const news$ = cheerio.load(news_response.data);

    const author: string = news$('.organisations')
      .first()
      .find('a')
      .text()
      .trim();
    if (!author) {
      console.log(`Cannot fetch author from page: ${url}`);
    }

    const content: string = news$(
      '.content.col.left.grid-col-8  > :not(:first-child)'
    )
      .text()
      .trim();
    if (!content) {
      console.log(`Cannot fetch content from page: ${url}`);
    }

    return { author, content };
  } catch (error) {
    console.log(`Cannot fetch page ${url}`);
    return { author: '', content: '' };
  }
}

export = gov_shared;
